File: android/app/src/main/java/com/ariafrittanative/queue/Queue.java
package com.ariafrittanative.queue;

import android.util.Log;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.modules.core.DeviceEventManagerModule;

import java.util.ArrayList;

public class Queue extends ReactContextBaseJavaModule {
    public static enum EVENT {
        REFRESH_TOKEN,
        RECIVE_NOTIFICATION;

        public String value() {
            switch (this) {
                case REFRESH_TOKEN:
                    return "onRefreshToken";
                case RECIVE_NOTIFICATION:
                    return "onReciveNotification";
                default:
                    return "defualt";
            }
        }
    }

    public static enum TYPE {
        PUSH_NOTIFICATION
    }

    private static ReactApplicationContext context = null;
    private static boolean jsReady = false;
    private static ArrayList<Element> queue = new ArrayList<>();
    private static ArrayList<TYPE> eventTypeEnabled = new ArrayList<>();

    Queue(ReactApplicationContext reactContext) {
        super(reactContext);
        context = reactContext;
    }

    private static class Element {
        final EVENT event;
        final TYPE type;
        final String detail;

        Element(EVENT event, TYPE type, String detail) {
            this.event = event;
            this.type = type;
            this.detail = detail;
        }
    }

    public static void enable(TYPE event) {
        eventTypeEnabled.add(event);

        jsReady = true;
        pop();
    }

    @Override
    public String getName() {
        return "Queue";
    }

    public synchronized static void push(EVENT event, TYPE type, String detail) {
        queue.add(new Element(event, type, detail));
        pop();
    }

    private synchronized static void pop() {
        if (jsReady && context != null) {
            ArrayList<Element> elementsToRemove = new ArrayList<>();
            for (Element e: queue) {
                if (eventTypeEnabled.contains(e.type)) {
                    context.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                            .emit(e.event.value(), e.detail);
                    Log.d("QUEUE", "pop - event: " + e.event.value() + ", value: " + e.detail);
                    // non ho potuto fare il remove direttamente qua perchè all'avvio dell'app in alcune casistiche crashava
                    elementsToRemove.add(e);
                }
            }
            queue.removeAll(elementsToRemove);
        }
    }
}

File: ios/AriaFrittaNative/Queue/Queue.m
//
//  Queue.m
//  AriaFrittaNative
//
//  Created by Giacomo Materozzi on 16/03/2020.
//  Copyright © 2020 Facebook. All rights reserved.
//

#import "Queue.h"

#pragma Event
@implementation Event {
@public NSString *event, *detail;
@public NSUInteger *type;
}

- (id) init: (NSString *) mEvent type: (NSUInteger *) mType detail:(NSString *) mDetail {
  if (self = [super init]) {
    event = mEvent;
    detail = mDetail;
    type = mType;
  }
  return self;
}

@end

#pragma QUEUE
@implementation Queue

NSMutableArray <Event *> *queue = nil;
NSMutableArray <NSNumber *> *eventEnabled = nil;

bool jsReady = false;
static Queue *sharedInstance = nil;

+ (id) allocWithZone:(NSZone *) zone {
  if (sharedInstance == nil) {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
      sharedInstance = [super allocWithZone:zone];
    });
  }
  return sharedInstance;
}

- (NSArray<NSString *> *) supportedEvents {
  return @[@"onRefreshToken", @"onReciveNotification"];
}

- (void) enable: (NSUInteger *) type {
  if (eventEnabled == nil) {
    eventEnabled = [NSMutableArray array];
  }
  
  [eventEnabled addObject:[NSNumber numberWithInt: type]];
  jsReady = true;
  [self pop];
}

- (void) push: (EVENT) event type: (NSUInteger *) type detail:(NSString * ) detail {
  @synchronized(self) {
    if (queue == nil) {
      queue = [NSMutableArray array];
    }
    
    NSString *eventName = @"default";
    
    if (event == REFRESH_TOKEN) {
      eventName = @"onRefreshToken";
    } else if (event == RECIVE_NOTIFICATION) {
      eventName = @"onReciveNotification";
    }
    
    Event *eventObject = [[Event alloc] init: eventName type:type detail: detail];
    [queue addObject:eventObject];
    [self pop];
  }
}

- (void) pop {
  @synchronized(self) {
    if (jsReady && self.bridge && queue != nil) {
      NSMutableArray *toDelete = [NSMutableArray array];
      for (Event *e in queue) {
        if ([eventEnabled containsObject: [NSNumber numberWithInt:e->type]]) {
          NSLog(@"%@", [@"call.. " stringByAppendingString: e->event]);
          [self sendEventWithName:e->event body: e->detail];
          // non ho potuto fare il remove direttamente qua perchè all'avvio dell'app in alcune casistiche crashava
          [toDelete addObject:e];
        } else {
          NSLog(@"%@ non avviato perchè il tipo: %lu non è stato abilitato", e->event, e->type);
        }
      }
      [queue removeObjectsInArray:toDelete];
    }
  }
}

RCT_EXPORT_MODULE();

@end
